File: 210404/semain_N-Queen.py
"""
python3로는 백트래킹 시간 초과로 통과 불가능.
pypy3만 가능.
"""

def check(row):
    for i in range(row):
        if chess[i]==chess[row] or abs(chess[row]-chess[i])==row-i:
            return False
    return True

def nqueen(row):
    global cnt
  
    if row == n:
        cnt+=1
        return
      
    for i in range(n):
        chess[row] = i
        if check(row):
            nqueen(row+1)

n = int(input())
chess = [0 for _ in range(16)]
cnt = 0
nqueen(0)
print(cnt)

File: 210404/semain_N-Queen.c
#include <stdio.h>
#include <stdlib.h>

int n;
int cnt = 0;
int chess[15];

int check(int line){
    for(int i=0; i<line; i++){
        if(chess[i]==chess[line] || abs(chess[line]-chess[i]) == line - i)
          return 0;
    }
    return 1;
}

void nqueen(int line){
    if (line==n){
        cnt++;
        return;
    }
    for(int i=0; i<n; i++){
        chess[line] = i;
        if(check(line)){
            nqueen(line + 1);
        }
    }
}

int main() {
    scanf("%d", &n);
    nqueen(0);
    printf("%d", cnt);
    return 0;
}
